lib/Core/Executor.cpp

```diff
@@ -1922,15 +1922,10 @@ void Executor::executeInstruction(ExecutionState &state, KInstruction *ki) {
   case Instruction::Select: {
     assert(i->getOperand(0)->getType()->isIntegerTy() &&
            "Invalid operand type");
-    // FIXME: Make the asserts for operand 1 and 2 stronger. At the moment all
-    // we really care about is making sure the operands are not of vector type.
-    // However for consistency with the rest of the code we should figure out
-    // which types are legal and assert that the operands are one of those
-    // types.
-    assert(!(i->getOperand(1)->getType()->isVectorTy()) &&
-           "Invalid operand type");
-    assert(!(i->getOperand(2)->getType()->isVectorTy()) &&
-           "Invalid operand type");
+    assert((i->getOperand(1)->getType() == i->getOperand(2)->getType()) &&
+           "true and false operands to Select have mismatching types");
+    // NOTE: We are do not check that operands 1 and 2 are not of vector type
+    // because the scalarizer pass might not remove these.
```

Comment:
Because you can have a select (if-then-else) instruction that picks from two different vector operands. The result of the select might then be consumed by an `insertelement` or `extractelement` instruction. One of the tests I wrote exposed this so I changed the assert.